### ISSUE TITLE
Identities should report an error

### ISSUE THREAD

**Issue by kkmuffme** [microsoft/TypeScript#55699]
### 🔍 Search Terms

Identity TS2367 impossible comparison

### ✅ Viability Checklist

- [X] This wouldn't be a breaking change in existing TypeScript/JavaScript code
- [X] This wouldn't change the runtime behavior of existing JavaScript code
- [X] This could be implemented without emitting different JS based on the types of the expressions
- [X] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
- [X] This feature would agree with the rest of our Design Goals: https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals

### ⭐ Suggestion

Comparing identical **literal** (abc, 5, false) values in an "if" it should report it as redundant/impossible, similar to what TS2367 does.

### 📃 Motivating Example

https://www.typescriptlang.org/play?#code/FAMwrgdgxgLglgewgAhAhAKAlALlQQwBsBnAU2QG9hkbkAnUmMOlEIsgbmAF9hRJYiFACN8dbHgDk+SZWq0GTFskmSuvftHhJkMAIwYxAczyQA1hAQB3CFjm1kcEMgyp02ZAEIAvN910wcjsqeRoNBycXN0w7Xz8YAKC5UORw2kjXNBivXwISJJCHNJoM6I84vLJkYL4i2vTnV1FxOx8-SWFZGpTix0bkZvLcjq7kupTSwdbhmWqx2l7JsSH22e663iA

```ts

function foo(): false {
    return false;
}

function bar(): 'a' {
    return '';
}

function t1(arg: unknown) {
    if ( foo() !== true ) {

    }

    if ( foo() === true ) {

    }

    if ( foo() !== false ) {

    }

    if ( foo() === false ) {

    }

    if ( bar() !== 'b' ) {

    }

    if ( bar() === 'b' ) {

    }

    if ( bar() !== 'a' ) {

    }

    if ( bar() === 'a' ) {

    }
}
```

### 💻 Use Cases

1. What do you want to use this for?
Get rid of legacy conditions/impossible code

2. What shortcomings exist with current approaches?
No error reported

4. What workarounds are you using in the meantime?
None, wasting time manually checking legacy code to see what can be removed.

**Comment by jcalz** [comment]
>- [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code

I imagine it would break existing code to issue new errors for comparisons that are currently allowed.  That might be acceptable, but you should acknowledge it.

**Comment by jcalz** [comment]
No, that's pretty much the definition of a breaking change.  If you check out [prior breaking changes](https://github.com/microsoft/TypeScript/wiki/Breaking-Changes) or [issues marked as "Breaking Change"](https://github.com/microsoft/TypeScript/issues?q=label%3A%22Breaking+Change%22+sort%3Acomments-desc) you will see that many are describing things that introduce new compile errors.

**Comment by Josh-Cena** [comment]
You want [`@typescript-eslint/no-unnecessary-condition`](https://typescript-eslint.io/play#ts=5.1.6&fileType=.tsx&code=CYUwxgNghgTiAEYD2A7AzgF3lAXPARAGZJL4DcAUBQJaHwAUU8AvKwcaQJTwDeF88CgF8KQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false). Beware that this has a lot of edge cases. For example, how do you know if you are doing an exhaustiveness check:

```js
if (a === "this") {}
else if (a === "that") {}
else {
  throw new Error("Non-exhaustive");
}
```

**Comment by RyanCavanaugh** [comment]
I don't see how this gets off the ground without a ton of special cases. For example, this code is 100.0% idiomatic:
```ts
Debug.assert(var === "value", "Got the expected value");
```

There's also the problem of arrays with `noUncheckedIndexedAccess` off:
```ts
function foo(arr: Array<"foo" | "bar">, idx: number) {
  const v= arr[idx;
  if (v === "foo") {
  } else if (v === "bar") { // <- is doing something useful!
  } else {
    // out of bounds; this code is reachable
  }
}
```
